packages/yew-test-runner/src/procedural.rs

```diff
@@ -0,0 +1,306 @@
+///
+/// Test properties and inspect the dom while you have one of these in sope.
+/// Borrows the context, so that you don't accidentally override the dom state with another test.
+pub struct TestableState<'s> {
+    context: &'s mut dyn TestContext,
+}
+
+impl<'s> fmt::Debug for TestableState<'s> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_struct("TestableState").finish_non_exhaustive()
+    }
+}
+
+impl<'s> TestableState<'s> {
+    /// Retrieve the DOM element under which the test has been mounted.
+    ///
+    /// Allows directly interacting with DOM API.
+    pub fn dom_root(&self) -> Element {
+        self.context.as_runner().parent.clone()
+    }
+
+    /// Test against an exactly given inner html that is supposedly rendered.
+    #[track_caller]
+    pub fn assert_inner_html(&self, expected: &'static str) {
+        let runner = self.context.as_runner();
+        let inner_html = runner.parent.inner_html();
+        assert_eq!(inner_html, expected, "inner html should match");
+    }
+}
+
+#[cfg(feature = "hydration")]
+mod feat_ssr_hydrate {
+    pub(super) use std::future::Future;
+    use std::ops::{Deref, DerefMut};
+    pub(super) use std::pin::Pin;
+
+    use super::*;
+
+    /// Access to the dom state, just before hydration occurs.
+    pub struct HydratableState<'s> {
+        pub(super) state: TestableState<'s>,
+        pub(super) test_case: Html,
+    }
+
+    impl fmt::Debug for HydratableState<'_> {
+        fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+            f.debug_struct("HydratableState").finish_non_exhaustive()
+        }
+    }
+
+    impl<'s> Deref for HydratableState<'s> {
+        type Target = TestableState<'s>;
+
+        fn deref(&self) -> &Self::Target {
+            &self.state
+        }
+    }
+
+    impl<'s> DerefMut for HydratableState<'s> {
+        fn deref_mut(&mut self) -> &mut Self::Target {
+            &mut self.state
+        }
+    }
+
+    impl<'s> HydratableState<'s> {
+        /// Hydrate the application now, and wait for it to render
+        pub fn hydrate(self) -> TestableState<'s> {
+            let runner = self.state.context.as_runner_mut();
+            // Hydrate the component into place
+            runner.test_app = Some(
+                Renderer::with_root_and_props(
+                    runner.parent.clone(),
+                    TestScaffoldProps {
+                        test_case: self.test_case,
+                    },
+                )
+                .hydrate(),
+            );
+            // Flush lifecycles, then return
+            yew::scheduler::__unstable_start_now();
+            TestableState {
+                context: self.state.context,
+            }
+        }
+    }
+}
+#[cfg(feature = "hydration")]
+pub use feat_ssr_hydrate::*;
+
+/// Common methods to [`TestRunner`] and [`TestStep`]. Automatically implemented for any type
+/// implementing [`TestContext`].
```

Comment:
```suggestion
/// Common methods to [`TestRunner`] and [`TestStep`]. 
///
/// Automatically implemented for any type implementing [`TestContext`].
```
Just formatting. This helps rustdoc only display the important line on the home page